src/ray/core_worker/pubsub/pubsub_coordinator.h

```diff
@@ -0,0 +1,163 @@
+  /// connection. False is used only for testing.
+  void QueueMessage(const ObjectID &object_id, bool try_publish = true);
+
+  /// Publish all queued messages if possible.
+  ///
+  /// \param force If true, we publish to the subscriber although there's no queued
+  /// message.
+  /// \return True if it publishes. False otherwise.
+  bool PublishIfPossible(bool force = false);
+
+ private:
+  /// Cached long polling reply callback.
+  LongPollConnectCallback long_polling_reply_callback_ = nullptr;
+  /// Queued messages to publish.
+  std::vector<ObjectID> mailbox_;
+};
+
+/// Pubsub server.
+class PubsubCoordinator {
+ public:
+  /// Pubsub coordinator constructor.
+  ///
+  /// \param is_node_dead A callback that returns true if the given node id is dead.
+  explicit PubsubCoordinator(std::function<bool(const NodeID &)> is_node_dead)
+      : is_node_dead_(is_node_dead) {}
+
+  ~PubsubCoordinator() = default;
+
+  ///
+  // TODO(sang): Currently, we need to pass the callback for connection because we are
+  // using long polling internally. This should be changed once the bidirectional grpc
+  // streaming is supported.
+  void Connect(const NodeID &subscriber_node_id,
+               LongPollConnectCallback long_poll_connect_callback);
+
+  /// Register the subscription.
+  ///
+  /// \param subscriber_node_id The node id of the subscriber.
+  /// \param object_id The object id that the subscriber is subscribing to.
+  void RegisterSubscription(const NodeID &subscriber_node_id, const ObjectID &object_id);
+
+  /// Publish the given object id to subscribers.
+  ///
+  /// \param object_id The object id to publish to subscribers.
+  void Publish(const ObjectID &object_id);
+
+  /// Remove the subscriber. Once the subscriber is removed, messages won't be published
+  /// to it anymore.
+  /// TODO(sang): Currently, clients don't send a RPC to unregister themselves.
+  ///
+  /// \param subscriber_node_id The node id of the subscriber to unsubscribe.
+  /// \return True if erased. False otherwise.
+  bool UnregisterSubscriber(const NodeID &subscriber_node_id);
+
+  /// Unregister subscription. It means the given object id won't be published to the
+  /// subscriber anymore.
+  ///
+  /// \param subscriber_node_id The node id of the subscriber.
+  /// \param object_id The object id of the subscriber.
+  /// \return True if erased. False otherwise.
+  bool UnregisterSubscription(const NodeID &subscriber_node_id,
+                              const ObjectID &object_id);
+
+ private:
+  /// Protects below fields. Since the coordinator runs in a core worker, it should be
+  /// thread safe.
+  mutable absl::Mutex mutex_;
+
+  /// Callback that returns true if the given node is dead.
+  std::function<bool(const NodeID &)> is_node_dead_;
+
+  /// Mapping of node id -> subscribers.
+  absl::flat_hash_map<NodeID, std::shared_ptr<Subscriber>> subscribers_
+      GUARDED_BY(mutex_);
```

Comment:
Are we testing that these data structures are all cleaned up with no leaks at the end of tests?